### PR TITLE
PSA Storage: Add psa_trusted_storage_linux persistent storage support for v1.0.0 APIs

### PR DESCRIPTION
**Pull request by simonqhughes**
This PR should be squashed and merged. The following is offered as the squash commit message:

```
Add psa_trusted_storage_linux persistent storage support for v1.0.0 APIs

The following provides more information on this PR:
- PSA stands for Platform Security Architecture.
- Add support for use of psa_trusted_storage_api internal-trusted_storage.h v1.0.0
  as the interface to the psa_trusted_storage_linux backend (i.e. for persistent
  storage when MBEDTLS_PSA_ITS_FILE_C is not defined). This requires changes
  to psa_crypto_its.h and psa_crypto_storage.c to migrate to the new API.
- Add MBEDTLS_PSA_ITS_FILE_C_INCLUDE_TEST to config.h to build
  useful storage related tests (e.g. test_suite_psa_its and 
  test_suite_psa_crypto_persistent_key) even when MBEDTLS_PSA_ITS_FILE_C 
  is not defined. This is used by the psa_trusted_storage_linux
  mbed_crypto_psa_storage_config.h for example.
- Add MBEDTLS_PSA_ITS_FILE_C_STORAGE_PREFIX in config.h so the
  file storage directory can be specified. It can also be specified in the
  following way:
      make CFLAGS="-DMBEDTLS_PSA_ITS_FILE_C_STORAGE_PREFIX='\"/home/username/\"'"
  (Note the inclusion of the trailing '/' in the prefix path.)
- Add CMake option for explicitly link library to trusted_storage
    option name: LINK_WITH_TRUSTED_STORAGE
    default value: ON
```

### CODE REVIEW

**Comment by simonqhughes** [comment]
@jenia81 @moshe-shahar @shelib01 Please request the project administrators to add you as reviewers asn I'm unable to do so.

File: library/CMakeLists.txt [hunk 1]
```diff
@@ -1,7 +1,6 @@
 option(USE_STATIC_MBEDTLS_LIBRARY "Build mbed TLS static library." ON)
 option(USE_SHARED_MBEDTLS_LIBRARY "Build mbed TLS shared library." OFF)
 option(LINK_WITH_PTHREAD "Explicitly link mbed TLS library to pthread." OFF)
-option(LINK_WITH_TRUSTED_STORAGE "Explicitly link mbed TLS library to trusted_storage." ON)
```
Comment (moshe-shahar) [inline library/CMakeLists.txt]: https://github.com/ARMmbed/mbed-crypto/pull/180/commits/5c6179145a09fbb8097b432a50c46030854063e9
@Patater ,
@simonqhughes revert this commit because of the tests fails.
Can you please help me integrate that change so it will not fails the tests?

**Comment by simonqhughes** [comment]
This PR has been submitted to mbedtls, which @Patater request for completing automated testing of the current PR:
https://github.com/ARMmbed/mbedtls/pull/2750

File: library/psa_its_file.c [hunk 2]
```diff
@@ -137,7 +137,8 @@ psa_status_t psa_its_get_info( psa_storage_uid_t uid,
 psa_status_t psa_its_get( psa_storage_uid_t uid,
                           uint32_t data_offset,
                           uint32_t data_length,
-                          void *p_data )
+                          void *p_data,
+                          size_t* p_data_length )
```
Comment (Patater) [inline library/psa_its_file.c]: Please put the `*` next to the variable name, not the type.

File: tests/scripts/all.sh [hunk 3]
```diff
@@ -726,6 +726,7 @@ component_test_no_platform () {
     scripts/config.pl unset MBEDTLS_FS_IO
     scripts/config.pl unset MBEDTLS_PSA_CRYPTO_STORAGE_C
     scripts/config.pl unset MBEDTLS_PSA_ITS_FILE_C
+    scripts/config.pl unset MBEDTLS_PSA_ITS_FILE_C_INCLUDE_TEST
```
Comment (Patater) [inline tests/scripts/all.sh]: We shouldn't need to add these lines when the `MBEDTLS_PSA_ITS_FILE_C_INCLUDE_TEST` option is removed.

File: tests/suites/test_suite_psa_crypto_persistent_key.data [hunk 4]
```diff
@@ -18,7 +18,7 @@ parse_storage_data_check:"505341004b4559":"":PSA_KEY_TYPE_RSA_KEYPAIR:PSA_KEY_US
 
 # Not specific to files, but only run this test in an environment where the maximum size could be reached.
 Save maximum size persistent raw key
-depends_on:MBEDTLS_PSA_ITS_FILE_C
+depends_on:MBEDTLS_PSA_ITS_FILE_C_INCLUDE_TEST
```
Comment (Patater) [inline tests/suites/test_suite_psa_crypto_persistent_key.data]: Not needed change anymore

File: tests/suites/test_suite_psa_its.function [hunk 5]
```diff
@@ -77,7 +83,7 @@ void set_get_remove( int uid_arg, int flags_arg, data_t *data )
     PSA_ASSERT( psa_its_get_info( uid, &info ) );
     TEST_ASSERT( info.size == data->len );
     TEST_ASSERT( info.flags == flags );
-    PSA_ASSERT( psa_its_get( uid, 0, data->len, buffer ) );
+    PSA_ASSERT( psa_its_get( uid, 0, data->len, buffer, &ret_len ) );
```
Comment (Patater) [inline tests/suites/test_suite_psa_its.function]: We should probably assert that `ret_len` is equal to `data->len`.

File: tests/suites/test_suite_psa_its.function [hunk 6]
```diff
@@ -98,21 +104,23 @@ void set_overwrite( int uid_arg,
...
+    size_t ret_len = 0;
 
     ASSERT_ALLOC( buffer, MAX( data1->len, data2->len ) );
 
     PSA_ASSERT( psa_its_set_wrap( uid, data1->len, data1->x, flags1 ) );
     PSA_ASSERT( psa_its_get_info( uid, &info ) );
     TEST_ASSERT( info.size == data1->len );
     TEST_ASSERT( info.flags == flags1 );
-    PSA_ASSERT( psa_its_get( uid, 0, data1->len, buffer ) );
+    PSA_ASSERT( psa_its_get( uid, 0, data1->len, buffer, &ret_len ) );
```
Comment (Patater) [inline tests/suites/test_suite_psa_its.function]: We should probably assert that `ret_len` is equal to `data1->len`.

File: tests/suites/test_suite_psa_its.function [hunk 7]
```diff
@@ -98,21 +104,23 @@ void set_overwrite( int uid_arg,
...
     ASSERT_COMPARE( data1->x, data1->len, buffer, data1->len );
 
     PSA_ASSERT( psa_its_set_wrap( uid, data2->len, data2->x, flags2 ) );
     PSA_ASSERT( psa_its_get_info( uid, &info ) );
     TEST_ASSERT( info.size == data2->len );
     TEST_ASSERT( info.flags == flags2 );
-    PSA_ASSERT( psa_its_get( uid, 0, data2->len, buffer ) );
+    ret_len = 0;
+    PSA_ASSERT( psa_its_get( uid, 0, data2->len, buffer, &ret_len ) );
```
Comment (Patater) [inline tests/suites/test_suite_psa_its.function]: We should probably assert that `ret_len` is equal to `data2->len`.

File: tests/suites/test_suite_psa_its.function [hunk 8]
```diff
@@ -143,11 +152,11 @@ void set_multiple( int first_id, int count )
     {
         mbedtls_snprintf( stored, sizeof( stored ),
                           "Content of file 0x%08lx", (unsigned long) uid );
-        PSA_ASSERT( psa_its_get( uid, 0, sizeof( stored ), retrieved ) );
+        PSA_ASSERT( psa_its_get( uid, 0, sizeof( stored ), retrieved, &ret_len ) );
```
Comment (Patater) [inline tests/suites/test_suite_psa_its.function]: We should probably assert that `ret_len` is equal to `sizeof( stored )`.

File: tests/suites/test_suite_psa_its.function [hunk 9]
```diff
@@ -143,11 +152,11 @@ void set_multiple( int first_id, int count )
...
         ASSERT_COMPARE( retrieved, sizeof( stored ),
                         stored, sizeof( stored ) );
         PSA_ASSERT( psa_its_remove( uid ) );
-        TEST_ASSERT( psa_its_get( uid, 0, 0, NULL ) ==
+        TEST_ASSERT( psa_its_get( uid, 0, 0, NULL, NULL ) ==
```
Comment (Patater) [inline tests/suites/test_suite_psa_its.function]: Good to see that we are also testing the `NULL` `data_length` parameter case.

File: tests/suites/test_suite_psa_its.function [hunk 10]
```diff
@@ -190,14 +199,15 @@ void get_at( int uid_arg, data_t *data,
...
+    size_t ret_len = 0;
 
     ASSERT_ALLOC( buffer, length + 16 );
     trailer = buffer + length;
     memset( trailer, '-', 16 );
 
     PSA_ASSERT( psa_its_set_wrap( uid, data->len, data->x, 0 ) );
 
-    status = psa_its_get( uid, offset, length_arg, buffer );
+    status = psa_its_get( uid, offset, length_arg, buffer, &ret_len );
```
Comment (Patater) [inline tests/suites/test_suite_psa_its.function]: We should probably assert that `ret_len` is equal to `length` when `status == PSA_SUCCESS`

File: include/mbedtls/config.h [hunk 11]
```diff
@@ -1734,6 +1734,41 @@
...
+ * (note the appended "/").
+ *
+ * Module:  library/psa_its_file.c
+ *
+ * Requires: MBEDTLS_FS_IO
+ */
+#if ! defined(MBEDTLS_PSA_ITS_FILE_C_STORAGE_PREFIX)
+#define MBEDTLS_PSA_ITS_FILE_C_STORAGE_PREFIX ""
+#endif
+
```
[the remark this answers is not in the document]
Comment (simonqhughes) [reply include/mbedtls/config.h]: OK

File: library/psa_crypto_storage.c [hunk 12]
```diff
@@ -40,6 +40,7 @@
 #else /* Native ITS implementation */
 #include "psa/error.h"
 #include "psa/internal_trusted_storage.h"
+
```
[the remark this answers is not in the document]
Comment (simonqhughes) [reply library/psa_crypto_storage.c]: OK, extra line removed. Thanks.

File: library/psa_its_file.c [hunk 13]
```diff
@@ -44,7 +44,7 @@
 #include <stdio.h>
 #include <string.h>
 
-#define PSA_ITS_STORAGE_PREFIX ""
+#define PSA_ITS_STORAGE_PREFIX MBEDTLS_PSA_ITS_FILE_C_STORAGE_PREFIX
```
[the remark this answers is not in the document]
Comment (simonqhughes) [reply library/psa_its_file.c]: OK thanks.

File: tests/suites/test_suite_psa_its.function [hunk 14]
```diff
@@ -1,5 +1,10 @@
 /* BEGIN_HEADER */
+#if defined(MBEDTLS_PSA_ITS_FILE_C)
 #include "../library/psa_crypto_its.h"
+#else
+#include "psa/internal_trusted_storage.h"
+#endif
```
[the remark this answers is not in the document]
Comment (simonqhughes) [reply tests/suites/test_suite_psa_its.function]: OK. Done.

File: library/psa_its_file.c (hunk 2 above)
Comment (simonqhughes) [reply library/psa_its_file.c]: Done.

File: tests/suites/test_suite_psa_its.function (hunk 5 above)
Comment (simonqhughes) [reply tests/suites/test_suite_psa_its.function]: Done

File: tests/suites/test_suite_psa_its.function (hunk 6 above)
Comment (simonqhughes) [reply tests/suites/test_suite_psa_its.function]: Done

File: tests/suites/test_suite_psa_its.function (hunk 7 above)
Comment (simonqhughes) [reply tests/suites/test_suite_psa_its.function]: Done

File: tests/suites/test_suite_psa_its.function (hunk 8 above)
Comment (simonqhughes) [reply tests/suites/test_suite_psa_its.function]: Done

File: tests/suites/test_suite_psa_its.function (hunk 9 above)
Comment (simonqhughes) [reply tests/suites/test_suite_psa_its.function]: Ok

File: tests/suites/test_suite_psa_its.function (hunk 10 above)
Comment (simonqhughes) [reply tests/suites/test_suite_psa_its.function]: Done

File: tests/scripts/all.sh (hunk 3 above)
Comment (simonqhughes) [reply tests/scripts/all.sh]: Done

File: tests/suites/test_suite_psa_crypto_persistent_key.data (hunk 4 above)
Comment (simonqhughes) [reply tests/suites/test_suite_psa_crypto_persistent_key.data]: OK, removed.

**Comment by simonqhughes** [comment]
This PR should be squashed and merged. The following is offered as the squash commit message having done the rework:

```
Add psa_trusted_storage_linux persistent storage support for v1.0.0 APIs

The following provides more information on this PR:
- PSA stands for Platform Security Architecture.
- Add support for use of psa_trusted_storage_api internal-trusted_storage.h v1.0.0
  as the interface to the psa_trusted_storage_linux backend (i.e. for persistent
  storage when MBEDTLS_PSA_ITS_FILE_C is not defined). This requires changes
  to psa_crypto_its.h and psa_crypto_storage.c to migrate to the new API.
```

File: library/psa_crypto_its.h [hunk 15]
```diff
@@ -106,7 +107,8 @@ psa_status_t psa_its_set(psa_storage_uid_t uid,
 psa_status_t psa_its_get(psa_storage_uid_t uid,
                          uint32_t data_offset,
                          uint32_t data_length,
-                         void *p_data);
+                         void *p_data,
+                         size_t* p_data_length );
```
Comment (Patater) [inline library/psa_crypto_its.h]: Please put the `*` next to the variable name.

File: library/psa_crypto_storage.c [hunk 16]
```diff
@@ -96,12 +96,13 @@ static psa_status_t psa_crypto_storage_load( const psa_key_file_id_t key,
...
     psa_storage_uid_t data_identifier = psa_its_identifier_of_slot( key );
     struct psa_storage_info_t data_identifier_info;
+    size_t data_length = 0;
 
     status = psa_its_get_info( data_identifier, &data_identifier_info );
     if( status  != PSA_SUCCESS )
         return( status );
 
-    status = psa_its_get( data_identifier, 0, (uint32_t) data_size, data );
+    status = psa_its_get( data_identifier, 0, (uint32_t) data_size, data, &data_length );
```
Comment (Patater) [inline library/psa_crypto_storage.c]: Should we check `data_length` is equal to `data_size` and return an error (probably `PSA_ERROR_STORAGE_FAILURE`) if not?

File: tests/suites/test_suite_psa_its.function [hunk 17]
```diff
@@ -190,18 +198,22 @@ void get_at( int uid_arg, data_t *data,
...
 
     PSA_ASSERT( psa_its_set_wrap( uid, data->len, data->x, 0 ) );
 
-    status = psa_its_get( uid, offset, length_arg, buffer );
+    status = psa_its_get( uid, offset, length_arg, buffer, &ret_len );
     TEST_ASSERT( status == (psa_status_t) expected_status );
     if( status == PSA_SUCCESS )
+    {
+        TEST_ASSERT( (size_t) length_arg == ret_len );
         ASSERT_COMPARE( data->x + offset, length,
```
Comment (Patater) [inline tests/suites/test_suite_psa_its.function]: We can do this in one assert if we replace the second `length` with `length_arg`. Same applies to other similar asserts added in this PR.

Comment (simonqhughes) [reply tests/suites/test_suite_psa_its.function]: I think you mean the code changes to the following (which uses ret_len rather than length_arg): 

```
    if( status == PSA_SUCCESS )
        ASSERT_COMPARE( data->x + offset, length,
                        buffer, ret_len );
```

File: library/psa_crypto_storage.c (hunk 16 above)
Comment (simonqhughes) [reply library/psa_crypto_storage.c]: yes, its better.

File: library/psa_crypto_its.h (hunk 15 above)
Comment (simonqhughes) [reply library/psa_crypto_its.h]: Done.

File: tests/suites/test_suite_psa_its.function (hunk 17 above)
Comment (Patater) [reply tests/suites/test_suite_psa_its.function]: Yes, I like your suggestion better, to match what's passed to `psa_its_get()`.

Comment (simonqhughes) [reply tests/suites/test_suite_psa_its.function]: Ok, will change it to the new version.

**Comment by Patater** [comment]
CI failure on merge run is flaky `m->m dtls1_2,yes TLS-ECDHE-RSA-WITH-AES-128-GCM-SHA256` test, which passed fine on the head run.
